src/wagtail_vector_index/storage/django.py

```diff
@@ -0,0 +1,602 @@
+        super().__init__(*args, **kwargs)
+
+
+class EmbeddableFieldsMixin(models.Model):
+    """Mixin for Django models that allows the user to specify which fields should be used to generate embeddings."""
+
+    embedding_fields = []
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def _get_embedding_fields(cls) -> list["EmbeddingField"]:
+        embedding_fields = {
+            (type(field), field.field_name): field for field in cls.embedding_fields
+        }
+        return list(embedding_fields.values())
+
+    @classmethod
+    def check(cls, **kwargs):
+        """Extend model checks to include validation of embedding_fields in the
+        same way that Wagtail's Indexed class does it."""
+        errors = super().check(**kwargs)
+        errors.extend(cls._check_embedding_fields(**kwargs))
+        return errors
+
+    @classmethod
+    def _has_field(cls, name):
+        try:
+            cls._meta.get_field(name)
+        except FieldDoesNotExist:
+            return hasattr(cls, name)
+        else:
+            return True
+
+    @classmethod
+    def _check_embedding_fields(cls, **kwargs):
+        errors = []
+        for field in cls._get_embedding_fields():
+            message = "{model}.embedding_fields contains non-existent field '{name}'"
+            if not cls._has_field(field.field_name):
+                errors.append(
+                    checks.Warning(
+                        message.format(model=cls.__name__, name=field.field_name),
+                        obj=cls,
+                        id="wagtailai.WA001",
+                    )
+                )
+        return errors
+
+
+class ModelFromDocumentOperator(FromDocumentOperator[models.Model]):
+    """A class that can convert Documents into model instances"""
+
+    def from_document(self, document: Document) -> models.Model:
+        # Use the first key in the list, which is the most specific model class
+        key = ModelKey(document.object_keys[0])
+        model_class = self._model_class_from_label(key.model_label)
+        try:
+            return model_class.objects.filter(pk=key.object_id).get()
+        except model_class.DoesNotExist as e:
+            raise IndexedTypeFromDocumentError("No object found for document") from e
+
+    def bulk_from_documents(
+        self, documents: Iterable[Document]
+    ) -> Generator[models.Model, None, None]:
+        documents = tuple(documents)
+
+        keys_by_model_label = self._get_keys_by_model_label(documents)
+        objects_by_key = self._get_models_by_key(keys_by_model_label)
+
+        yield from self._get_deduplicated_objects_generator(documents, objects_by_key)
+
+    async def abulk_from_documents(
+        self, documents: Iterable[Document]
+    ) -> AsyncGenerator[models.Model, None]:
+        """A copy of `bulk_from_documents`, but async"""
+        # Force evaluate generators to allow value to be reused
+        documents = tuple(documents)
```

Comment:
Yeah you're right - I've changed these methods to accept Sequences to prevent the need to force evaluate. You could still conceivably pass a Sequence that can't be evaluated in an async context but I'm not sure if there's a way to guard against that - any ideas welcome!

Reply:
Looks good! Honestly I can't think of a great way to impose that restriction either - ie get rid of the chance of `.only` etc deferring fields we want - without forcing it to no longer be a sequence of documents, and instead plain dataclasses... which seems like overkill.